### DISCUSSION TITLE
Restful API 介紹&範例

### DISCUSSION THREAD

**Discussion by fatsoap** [1101-GS3073-C/Python-Community-GS3073-C#291] [General]
## 前情提要

由於最近上課都在教怎麼用 flask ，相信大家對於 API 是什麼有一定的瞭解了！不過問題是 API 應該要如何設計，一個好的 API 設計方式最好是能夠依照制定好的規則去設計，這樣可以有效在大型專案中的整合各個 API 接口 :rocket:

## RESTful API 是什麼?

> **Representational State Transfer** 一種 API 的設計風格，讓對相同資源 的 API 操作在同一個 path ，用不同的 Http method 來描述動作。

→ 由 path 來決定要對甚麼資源操作

→ 由 Http Method 來決定要做什麼操作

→ Content-Type 可用 JSON , multipart form-data , x-www-form-urlencoded 等等

### Restful API 的設計風格分為三個部分

1. Url Path : 用來描述這個 path 的是對什麼物件進行操作，通常會用名詞

- 例如對用戶進行操作的 path 可以用 `user` ，對訂單可以用 `order`

```
# User API
https://domain/user

# Order API
https://domain/order
```

2. Method : 使用 Http 定義的 request method 來描述要做什麼行為

```
GET 拿物件 ( safe & idempotent )
POST 新增物件
PUT 新增、更新物件 ( idempotent )
PATCH 更新物件
DELETE 刪除物件 ( idempotent )
```

> Save

是指該操作不會改變原本物件資料，例如: 查看訂單是不會改變訂單本身紀錄。

> Idempotent

是指該操作不管做 1 遍、2 遍或多遍，都會得到同樣的物件資料結果，。例如: 同一筆資料被 DELETE 了 2 次，雖然都是一樣的結果。

> PUT

輸入整個物件的資料，包含 id 、有更新的資料、沒有更新的資料，所以資料庫沒有的話就會新增，有的話就會根據 id 去更新。

> PATCH

只輸入有更新的部分物件，例如只更新用戶的信箱，就只會出現輸入 email。

3. Content Types

常用 json 作為資料傳輸的種類，但是 multipart form-data , x-www-form-urlencoded 等等也可以

## Rest 設計風格的優點和特徵

1. 統一的 API 接口 Uniform Interface

- 有唯一的 URL 來表示指定資源的位置，不會有混亂的 API path 命名

2. 無狀態 Stateless

- 比起已往的 cookie & session 的方式，使用 token 夾在 request 中，不只能夠降低 server 端 memory 的使用，也因為每一次 request 都是獨立的，能夠設計成分散式架構，擁有很高的水平擴充能力(Scale Out)

3. 可快取 Cachable

- 藉由 cache 之前 GET 過的資料來提升回應速度

4. 分層系統架構 (Layered System)

- 請求會經過 Load Balancer 或是 Proxy ，Client 端不知道回應的伺服器是哪個，這樣有助於提升擴充性 (Scalability) & 安全性

5. 客戶端 & 伺服器分離 (Client-Server)

- 有助於提高使用者介面的跨平台的可移植性 & Server side 的 horizontal scalibility

### GET

```
# Get User By ID
GET : {{host}}/user/:id

# 取得 id 為 123 的用戶資料
GET : http://localhost/user/123
```

```
# Get User By Query
GET : {{host}}/users

# 取的權限是 admin 的５個用戶，跳過前10個
GET : http://localhost/users?limit=5&skip=10&role=admin
```

### POST

```
# Create User
POST : {{host}}/user
Content-Type: application/json

# 新增 name 是 ABC 、 phone 是 123 的用戶
POST : http://localhost/user
Content-Type: application/json
{
	"name": "ABC",
	"phone": "123"
}
```

### PUT

```
# Update or Create User
PUT : {{host}}/user/:id
Content-Type: application/json

# 更新 id 是 123 的用戶的 name 改成 ABC 、 phone 改成 789
PUT : http://localhost/user
Content-Type: application/json
{
	"id": 123
	"name": "ABC",
	"phone": "789"
}
```

### PATCH

```
# Update User By ID
PATCH : {{host}}/user/:id
Content-Type: application/json

# 更新 id 是 123 的用戶的 phone 改成 456
PATCH : http://localhost/user/123
Content-Type: application/json
{
	"phone": "456"
}
```

### DELETE

```
# Delete User By ID
DELETE : {{host}}/user/:id

# 刪除 id 是 123 的用戶
DELETE : http://localhost/user/123
```

![](https://media1.giphy.com/media/XtydbjSSwkC7K2zBTH/giphy.gif?cid=790b7611b0be664e5141aa53431ce9bcc3f6c30485376e5f&rid=giphy.gif&ct=g)

## References

[API 是什麼? RESTful API 又是什麼?](https://medium.com/itsems-frontend/api-%E6%98%AF%E4%BB%80%E9%BA%BC-restful-api-%E5%8F%88%E6%98%AF%E4%BB%80%E9%BA%BC-a001a85ab638)
[什麼是 REST/RESTful ?](https://ithelp.ithome.com.tw/articles/10157431)
[Postman 常見的 Content-type](https://medium.com/hobo-engineer/ricky%E7%AD%86%E8%A8%98-postman-%E5%B8%B8%E8%A6%8B%E7%9A%84-content-type-b17a75396668)
[表現層狀態轉換](https://zh.wikipedia.org/wiki/%E8%A1%A8%E7%8E%B0%E5%B1%82%E7%8A%B6%E6%80%81%E8%BD%AC%E6%8D%A2)
[Horizontal vs Vertical Scaling](https://touchstonesecurity.com/horizontal-vs-vertical-scaling-what-you-need-to-know/)

**Comment by enoch-lightning** [comment]
這些都是python語法嗎？還是什麼？

**Reply by enoch-lightning** [reply]
![giphy](https://user-images.githubusercontent.com/86505454/160276670-7386ac69-1e0e-4131-ba69-5c31c6bda887.gif)
